metadata

DominikWolek opened the pull request:
closes #51, closes #53, closes #56, closes #52, closes #57

DominikWolek committed:
Add basic extractor

DominikWolek committed:
Add metadata gc

DominikWolek committed:
Polish metadata gathering for presentation

DominikWolek committed:
Add File and Metadata managers

DominikWolek committed:
Simplify FileManager

DominikWolek committed:
Refactor MetadataManager

DominikWolek committed:
Add FileTextExtractor interface

Future implementations: TikaTextExtractor, some OCRTextExtractor

DominikWolek committed:
Cleanup Metadata code

Deleted TikaExtrator, MetadataCrawler and MetadataGarbageCollector

Fixed code format

Fixes imports

DominikWolek committed:
Change relation between FileThumbnail and FileMetadata

Thumbnail now stores path, not ByteArray

DominikWolek committed:
Remove unecessary FileThumbnailRepository

DominikWolek committed:
Add status to FileEntry

DominikWolek committed:
Delete unecessary functions from FileManager

kkiermasz commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/repositories/FileEntryRepository.kt:
```diff
@@ -1,6 +1,13 @@
 
-interface FileEntryRepository : JpaRepository<FileEntry, String>
+interface FileEntryRepository : JpaRepository<FileEntry, String> {
+    fun existsByPath(path: String): Boolean?
```
Code style consistency violation. Add empty lines between methods.

kkiermasz replied:
`existsByPath(path: String)` should return non-optional `Boolean`

kkiermasz commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/repositories/FileContentRepository.kt:
```diff
@@ -0,0 +1,11 @@
+
+interface FileContentRepository : JpaRepository<FileContent, Long> {
+    @Transactional
+    fun deleteByFile(id: FileMetadata?)
```
I'm not sure what that method does. Consider renaming it

kkiermasz commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/managers/MetadataManager.kt:
```diff
@@ -0,0 +1,21 @@
+interface MetadataManager {
+    fun extractAndSaveMetadata(file: File): FileMetadata
+
+    fun getAllEntries(metadata: FileMetadata): Set<FileEntry>
```
If you return Array instead of Set, client will not be forced to take care of ordering the results.

DominikWolek replied:
Array does not provide unique results. 

Why should the results be ordered?

kkiermasz replied:
Because client will display these entries. 
Taking care of ordering them by default would be easy to achieve by this manager, because of how the database works.

 
Take a look at this: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/distinct-by.html

kkiermasz commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/managers/FileManager.kt:
```diff
@@ -0,0 +1,18 @@
+
+    fun rejectFile(path: String)
+
+    fun removeFile(path: String)
```
```suggestion
    fun removeEntry(path: String)
```
I suppose this method will delete both file and dir

kkiermasz commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/entities/FileMetadata.kt:
```diff
@@ -3,17 +3,16 @@ package pl.edu.uj.ii.ksi.mordor.persistence.entities
 data class FileMetadata(
     @Id
-    @GeneratedValue(strategy = GenerationType.AUTO)
+    @Column(name = "metadata_id")
```
Do you have to call it `metadata_id` since it is nested in FileMetada?

kkiermasz commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/managers/MetadataManager.kt:
```diff
@@ -0,0 +1,21 @@
+import pl.edu.uj.ii.ksi.mordor.persistence.entities.FileThumbnail
+
+interface MetadataManager {
+    fun extractAndSaveMetadata(file: File): FileMetadata
```
SRP
Client doesn't need to know what happens inside. I suggest renaming it to `saveFileMetadata` and do not return it.

DominikWolek replied:
How will the client find this specific metadata later, i.e. to create new FileEntry?

kkiermasz reviewed (changes requested):
Consider nesting files you've created. Especially managers.

DominikWolek committed:
Delete unecessary FileEntryRepository

Change existsByPath return type to Boolean

Rename removeFile to removeEntry

Delete unecessary specification for column name in FileMetadata

apardyl commented on build.gradle:
```diff
@@ -70,6 +70,7 @@ dependencies {
     compile("org.jetbrains.kotlin:kotlin-reflect")
     compile("org.glassfish.jaxb:jaxb-runtime:2.3.2")
     compile("commons-io:commons-io:2.6")
+    compile("org.apache.tika:tika-parsers:1.11")
```
1.11 release from 2015? latest is 1.24.

DominikWolek committed:
Remove getFileThumbnail from MetadataManager

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/entities/FileContent.kt:
```diff
@@ -1,18 +1,20 @@
 
-    @MapsId
-    @OneToOne(optional = false)
+    @OneToOne(optional = false, cascade = [CascadeType.DETACH])
```
you probably want cascade = CascadeType.ALL, orphanRemoval = true

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/entities/FileContent.kt:
```diff
@@ -1,18 +1,20 @@
+    @GeneratedValue(strategy = GenerationType.SEQUENCE)
     var id: Long? = null,
 
-    @MapsId
```
why was this replaced with GeneratedValue?

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/entities/FileMetadata.kt:
```diff
@@ -3,17 +3,15 @@ package pl.edu.uj.ii.ksi.mordor.persistence.entities
 data class FileMetadata(
     @Id
-    @GeneratedValue(strategy = GenerationType.AUTO)
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
why was that changed?

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/entities/FileMetadata.kt:
```diff
@@ -27,12 +25,6 @@ data class FileMetadata(
 
     var mimeType: String?,
 
-    @OneToOne(mappedBy = "file", orphanRemoval = true, cascade = [CascadeType.REMOVE], fetch = FetchType.LAZY)
```
why did you remove all reverse mappings?

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/entities/FileThumbnail.kt:
```diff
@@ -2,20 +2,21 @@ package pl.edu.uj.ii.ksi.mordor.persistence.entities
+    @GeneratedValue(strategy = GenerationType.SEQUENCE)
     var id: Long? = null,
 
-    @MapsId
```
as above

apardyl reviewed (changes requested):
Review part I.

kkiermasz reviewed (changes requested):
Please take a look at previous review. You did not return it to me, but changes were made.

DominikWolek committed:
Update Tika

DominikWolek committed:
Revert "Add status to FileEntry"

This reverts commit affa49e25c00627fa229e36eb27345b4d1198cc9.

DominikWolek committed:
Apply changes required by reviewer in File* entities

DominikWolek committed:
Add TikaFileTextExtractor

DominikWolek committed:
Add FileUploadSession

DominikWolek committed:
Refactor User creation to start with empty ArrayList of sessions

DominikWolek committed:
Update user-related tests

DominikWolek committed:
Add MetadaGarbageCollector

DominikWolek committed:
Change FileUploadSession from entitity to data class

DominikWolek committed:
Add MetadataCrawler

apardyl commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,5 +1,6 @@
+#Tue Apr 21 23:36:33 CEST 2020
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.10.2-all.zip
```
revert?

DominikWolek committed:
Finish FileEntryService

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/entities/FileEntry.kt:
```diff
@@ -12,4 +12,5 @@ data class FileEntry(
 
     @ManyToOne(fetch = FetchType.EAGER, optional = false)
     var metadata: FileMetadata
+
```
nit: revert

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/entities/User.kt:
```diff
@@ -31,4 +31,5 @@ data class User(
     @Enumerated(EnumType.STRING)
     @Column(nullable = false)
     var role: Role = Role.NOBODY
+
```
nit: revert

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/managers/FileManager.kt:
```diff
@@ -0,0 +1,18 @@
+
+    fun getFileEntry(path: String): FileEntry?
+
+    fun approveFile(path: String)
```
replace approveFile nad rejectFile with moveFile - this class need not be aware of file approval logic.

kkiermasz replied:
From now `FileUploadSessionService` is responsible for that.

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/managers/MetadataManager.kt:
```diff
@@ -0,0 +1,18 @@
+
+    fun removeMetadata(metadata: FileMetadata)
+
+    fun getFileContent(metadata: FileMetadata): FileContent
```
useless - FileMetadata.crawledContent returns the same

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/repositories/FileMetadataRepository.kt:
```diff
@@ -3,6 +3,4 @@ package pl.edu.uj.ii.ksi.mordor.persistence.repositories
 import pl.edu.uj.ii.ksi.mordor.persistence.entities.FileMetadata
 
-interface FileMetadataRepository : JpaRepository<FileMetadata, Long> {
-    fun findByFileHash(fileHash: String): FileMetadata?
```
why are you removing this?

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/repositories/FileThumbnailRepository.kt:
```diff
@@ -1,6 +0,0 @@
-import org.springframework.data.jpa.repository.JpaRepository
-import pl.edu.uj.ii.ksi.mordor.persistence.entities.FileThumbnail
-
-interface FileThumbnailRepository : JpaRepository<FileThumbnail, Long>
```
as above.

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/upload/session/FileUploadSessionService.kt:
```diff
@@ -0,0 +1,15 @@
+
+    fun reject(uploadSession: FileUploadSession) {}
+
+    fun getAllSessions(): List<FileUploadSession> {
```
make it accept a Pageable and return a Page - this will help with pagination in view.

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/upload/session/FileUploadSessionService.kt:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.stereotype.Service
+
+@Service
+class FileUploadSessionService {
```
Consider implementing a CrudRepository or PagingAndSortingRepository instead.

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/upload/session/FileUploadSession.kt:
```diff
@@ -0,0 +1,9 @@
+import pl.edu.uj.ii.ksi.mordor.persistence.entities.User
+
+data class FileUploadSession(
+    var id: String,
```
make it immutable - this is a DAO not a model.

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/upload/session/FileUploadSession.kt:
```diff
@@ -0,0 +1,9 @@
+data class FileUploadSession(
+    var id: String,
+
+    var user: User
```
as above

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/upload/session/FileUploadSession.kt:
```diff
@@ -0,0 +1,9 @@
+    var id: String,
+
+    var user: User
+)
```
add creationDate

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/managers/FileManager.kt:
```diff
@@ -0,0 +1,18 @@
+import java.io.File
+import pl.edu.uj.ii.ksi.mordor.persistence.entities.FileEntry
+
+interface FileManager {
```
How does this class relates to RepositoryService?

DominikWolek replied:
This class was supposed to be facade of FileEntry related services.

apardyl replied:
since FileEntries are secondary to actual files it looks like you are duplicating the functionality of RepositoryService.

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/repositories/FileEntryRepository.kt:
```diff
@@ -1,6 +1,13 @@
+    fun findAllByMetadata(metadata: FileMetadata): List<FileEntry>?
+
+    @Transactional
+    fun deleteByPath(path: String)
```
Duplicates deleteById?

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/entities/FileContent.kt:
```diff
@@ -12,7 +13,7 @@ data class FileContent(
 
     @MapsId
-    @OneToOne(optional = false)
+    @OneToOne(optional = false, cascade = [CascadeType.ALL], orphanRemoval = true)
```
I'm not sure on what side of the relation orphanRemoval should be set, please double check.

DominikWolek replied:
orphanRemoval should be placed in parent, so this should be removed

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/FileEntryService.kt:
```diff
@@ -0,0 +1,27 @@
+    private val entryRepository: FileEntryRepository
+) {
+    fun fileExists(entry: FileEntry): Boolean {
+        return repositoryService.getEntity(entry.path) != null
```
implement this in repositoryService to avoid building a full RepositoryEntity on each call.

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/FileEntryService.kt:
```diff
@@ -0,0 +1,27 @@
+        return repositoryService.getEntity(entry.path) != null
+    }
+
+    fun checkFile(entry: FileEntry) {
```
Move to MetadataGarbageCollector and make private.

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/FileEntryService.kt:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    fun isFileValidEntryCandidate(file: File): Boolean {
+        return file.exists() && !file.isDirectory && !file.isHidden
```
use isFile instead of !file.isDirectory (e.g. a unix socket is neither a dir nor a file). Also, maybe move it to RepositoryEntity (as needsMetadata(), where the default is false, and RepositoryFile overrides with an actual implementation)?

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/MetadataGarbageCollector.kt:
```diff
@@ -0,0 +1,33 @@
+        logger.info("Metadata garbage collection finished successfully")
+    }
+
+    private fun checkMetadata(metadata: FileMetadata?) {
```
pass it as FileMetadata (without `?`)

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/MetadataCrawler.kt:
```diff
@@ -0,0 +1,35 @@
+    @Value("\${mordor.root_path}") private val rootPathStr: String,
+    val fileEntryService: FileEntryService,
+    val fileEntryRepository: FileEntryRepository,
+    val fileEntryCreator: FileEntryCreator
```
I can't find a declaration for this class!!!!111oneone

DominikWolek committed:
Add getAllFiles to FileUploadSession

DominikWolek committed:
Add metadata extractor

DominikWolek committed:
Delete unecessary Metadata Manager

DominikWolek committed:
Add FileEntryCreator

DominikWolek committed:
Add FileContentRepository

DominikWolek committed:
Add findByFileHash to FileMetadataRepository

DominikWolek committed:
Delete unecessary argument from addFile

DominikWolek committed:
Revert "Update Tika" changes in grale-wrapper.properties

This reverts commit fe9735f17a9c253efdfe6c9e0e07f6994d563295.

DominikWolek committed:
Revert "Remove unecessary FileThumbnailRepository"

This reverts commit 6d186e2d4e0186d108737297d5a7ff835f069381.

DominikWolek committed:
Implement fileExists in RepositoryService

DominikWolek committed:
Remove orphanRemoval on metadata in FileContent

DominikWolek committed:
Delete FileEntryService

Move validation of file existance and need for metadata to RepositoryService or RepositoryEntities

DominikWolek committed:
Apply changes requeste by reviewer

DominikWolek committed:
Fix code formatting

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/repository/RepositoryFile.kt:
```diff
@@ -96,4 +96,7 @@ class RepositoryFile(
     }
+
+    override fun needsMetadata(): Boolean { return true }
+    //TODO: check for hidden files
```
also check if metadata is already available.

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/TikaMetadataExtractor.kt:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    private fun calculateHash(file: File): String? {
+        return DigestUtils.md5DigestAsHex(Files.readAllBytes(file.toPath()))
```
use sha256 instead https://en.wikipedia.org/wiki/MD5#Collision_vulnerabilities

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/FileEntryCreator.kt:
```diff
@@ -0,0 +1,54 @@
+    private val fileTextExtractor: FileTextExtractor
+) {
+
+    fun create(file: File): FileEntry {
```
rename to findOrCreate, try to first find a matching existing FileContent by hash, parse only if not found.

DominikWolek replied:
FileEntry should be created even if metadata with the same hash exists, as it only maps a path to metadata. Therefore I think `create` is a better name for this function

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/MetadataCrawler.kt:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    private fun needMetadata(file: File): Boolean {
+        return repositoryService.fileExists(file.path) && repositoryService.getEntity(file.path)?.needsMetadata()!!
```
```kotlin
return repositoryService.getEntity(file.path)?.needsMetadata() ?: false
```

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/persistence/entities/FileContent.kt:
```diff
@@ -12,7 +13,7 @@ data class FileContent(
     @MapsId
-    @OneToOne(optional = false)
+    @OneToOne(optional = false, cascade = [CascadeType.ALL])
     var file: FileMetadata?,
```
```kotlin
var files: List<FileEntry>? = mutableListOf()
```
this way you won't have to pass an empty list every time

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/FileEntryCreator.kt:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    private fun createContent(file: File, metadata: FileMetadata) {
+        val text = fileTextExtractor.extract(file)!!
```
either pass a nullable or non-null String, do not convert to non-null

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/FileEntryCreator.kt:
```diff
@@ -0,0 +1,54 @@
+    private fun createContent(file: File, metadata: FileMetadata) {
+        val text = fileTextExtractor.extract(file)!!
+        val content = FileContent(text = text, file = metadata)
+        contentRepository.save(content)
```
Throws an error `detached entity passed to persist`.
Also... have you actually tested this before committing? Like whether it actually works?

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/upload/session/FileUploadSessionService.kt:
```diff
@@ -0,0 +1,21 @@
+
+    fun reject(uploadSession: FileUploadSession) {}
+
+    fun getAllFiles(uploadSession: FileUploadSession): List<FileEntry> {
```
make this a method on FileUploadSession

DominikWolek replied:
`getAllFiles` should be implemented with use of the `RepositoryService`. Should such a method be part of dao?

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/repository/RepositoryFile.kt:
```diff
@@ -96,4 +96,7 @@ class RepositoryFile(
     }
+
+    override fun needsMetadata(): Boolean { return true }
+    // TODO: check for hidden files
```
also check if metadata is already present.

DominikWolek committed:
Remove FileManager - it is redundant to RepositoryService

DominikWolek committed:
Check fileHash before extracting metadata

DominikWolek committed:
Add FileHashProvider

DominikWolek committed:
Set default of FileMetadata.files to mutableList

DominikWolek committed:
Remove unused property in Metadata Crawler

DominikWolek committed:
Check if metadata exists on needMetadata in RepositoryFile

DominikWolek committed:
Move messageDigest inside ShaHashProvider

DominikWolek committed:
Add toString implementation to metadata entities

Implementation is based on ids

Lack of toString caused StackOverflowException on metadata creation

DominikWolek committed:
Change hash algorithm to SHA-256

DominikWolek committed:
RepositoryService returns metadata with RepositoryEntity

Since RepositoryService use FileEntryRepository to get metadata, tests
were updated to inject mocked FileEntryRepository

Possible future improvments: add thumbnail

DominikWolek committed:
Allow to set max length in FileTextExtractors

DominikWolek committed:
Fix FileEntryCreator

Entities are now saved in right order

DominikWolek committed:
Add note do crawler

DominikWolek committed:
WIP Change api of FileUploadSession to PagingAndSortingRepository

DominikWolek committed:
WIP: Add RepositoryService saveFile function

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/FileEntryCreator.kt:
```diff
@@ -0,0 +1,78 @@
+import pl.edu.uj.ii.ksi.mordor.services.hash.FileHashProvider
+
+@Service
+class FileEntryCreator(
```
You should first get all data and then save all to db in one transaction - see what happens now if e.g. saving thumbnail fails.

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/FileEntryCreator.kt:
```diff
@@ -0,0 +1,78 @@
+@Service
+class FileEntryCreator(
+    private val metadataExtractor: MetadataExtractor,
+    private val entryRepository: FileEntryRepository,
```
You might prefer to use an EntityManager when saving a lot of different models.

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/FileEntryCreator.kt:
```diff
@@ -0,0 +1,78 @@
+        private const val contentMaxLength: Int = 24 * 1024
+    }
+
+    fun create(file: File) {
```
Creating a shitton of tiny functions does NOT actually improve readability. Please apply common sense.

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/FileTextExtractor.kt:
```diff
@@ -0,0 +1,8 @@
+
+interface FileTextExtractor {
+    fun extract(file: File): String?
+    fun maxLength(maxLength: Int): FileTextExtractor
```
Either pass it as a parameter to extract (with a default value) or create a FileTextExtractorBuilder/Factory class if you really must.

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/TikaFileTextExtractor.kt:
```diff
@@ -0,0 +1,28 @@
+
+    private fun trim(text: String): String {
+        return if (maxLength != null) {
+            text.toCharArray().take(maxLength ?: text.length).toString()
```
WTF?
![images](https://user-images.githubusercontent.com/10090582/80093801-74716100-8565-11ea-919c-c091fda95ddc.jpeg)

DominikWolek committed:
Change return type of addFile to RepositoryEntity?

Change getAllFiles return type to list of RepositoryEntities

DominikWolek committed:
Implement FileUploadRepository

DominikWolek committed:
Improve metadata creation

DominikWolek committed:
Update tests

DominikWolek committed:
Delete unecessary import

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/FileEntryCreator.kt:
```diff
@@ -0,0 +1,76 @@
+import pl.edu.uj.ii.ksi.mordor.services.repository.RepositoryService
+
+@Service
+@Transactional
```
Use transaction only when actually working with DB

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/TikaFileTextExtractor.kt:
```diff
@@ -0,0 +1,31 @@
+            }
+        } catch (e: IOException) {
+            logger.warn("Extraction of file text failed, returning empty text instead", e)
+            ""
```
return null instead?

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/upload/session/FileUploadSessionService.kt:
```diff
@@ -0,0 +1,47 @@
+    private val repositoryService: RepositoryService,
+    @Value("\${mordor.pending_sessions_path:}") private val pendingSessionsPath: String
+) {
+    fun getAllFiles(uploadSession: FileUploadSession): List<RepositoryEntity> {
```
return RepositoryService instead (this way saveFileInSession can be removed)

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/upload/session/FileUploadSessionService.kt:
```diff
@@ -0,0 +1,47 @@
+        repositoryService.saveFile(path, streamSource)
+    }
+
+    fun createSingleFileSession(user: User, path: String, streamSource: InputStreamSource): FileUploadSession {
```
change to createFileSession and create an empty session (SRP)

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/upload/session/FileUploadSessionRepository.kt:
```diff
@@ -0,0 +1,126 @@
+    override fun <S : FileUploadSession> save(session: S): S {
+        val absolutePath = repositoryService.getAbsolutePath(getPathOfSession(session))
+        absolutePath.toFile().mkdirs()
+        // TODO: set timestamp
```
isn't timestamp set by OS?

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/repository/RepositoryService.kt:
```diff
@@ -33,6 +44,80 @@ class RepositoryService(@Value("\${mordor.root_path}") private val rootPathStr:
+        if (entity is RepositoryFile) {
+            Files.delete(absolutePath)
+            entryRepository.deleteById(absolutePath.toString())
+        } else {
```
if else entity is RepositoryDirectory

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/repository/RepositoryService.kt:
```diff
@@ -33,6 +44,80 @@ class RepositoryService(@Value("\${mordor.root_path}") private val rootPathStr:
+    }
+
+    @Transactional
+    fun delete(path: String) {
```
for safety add recursive: Bool = false param

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/repository/RepositoryService.kt:
```diff
@@ -33,6 +44,80 @@ class RepositoryService(@Value("\${mordor.root_path}") private val rootPathStr:
+    }
+
+    @Transactional
+    fun move(from: String, to: String) {
```
for safety add recursive: Bool = false param

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/repository/RepositoryService.kt:
```diff
@@ -33,6 +44,80 @@ class RepositoryService(@Value("\${mordor.root_path}") private val rootPathStr:
+        if (entity is RepositoryFile) {
+            Files.move(absoluteFromPath, absoluteToPath)
+            moveEntry(absoluteToPath, absoluteFromPath)
+        } else {
```
if else entity is RepositoryDirectory

(we might introduce a new entity one day)

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/MetadataGarbageCollector.kt:
```diff
@@ -0,0 +1,39 @@
+        entryRepository.findAllByMetadata(metadata)?.forEach { entry -> checkFile(entry) }
+    }
+
+    private fun checkFile(entry: FileEntry) {
```
log file name with debug level

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/MetadataCrawler.kt:
```diff
@@ -0,0 +1,37 @@
+        File(rootPathStr).walkTopDown()
+                .filter { file -> needMetadata(file) }
+                .forEach { file ->
+                    fileEntryCreator.create(file) }
```
log file name with debug level

DominikWolek committed:
Apply changes requested by reviewer

DominikWolek committed:
Fix formatting

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/repository/RepositoryService.kt:
```diff
@@ -33,6 +44,78 @@ class RepositoryService(@Value("\${mordor.root_path}") private val rootPathStr:
+        return getAbsolutePath(path).toFile().exists()
+    }
+
+    fun saveFile(path: String, inputStreamSource: InputStreamSource): RepositoryEntity? {
```
pass as InputStream not InputStreamSource

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/repository/RepositoryService.kt:
```diff
@@ -33,6 +44,78 @@ class RepositoryService(@Value("\${mordor.root_path}") private val rootPathStr:
+
+        outputPath.toFile().mkdirs()
+        val outputStream = FileOutputStream(outputPath.toString())
+        outputStream.write(inputStream.readBytes())
```
use `copyTo` instead - readBytes loads entire file to memory.

apardyl commented on src/main/kotlin/pl/edu/uj/ii/ksi/mordor/services/FileEntryCreator.kt:
```diff
@@ -0,0 +1,75 @@
+        return entry
+    }
+
+    private fun findMetadataWithSameHash(file: File): FileMetadata? {
```
use repository here?

DominikWolek committed:
Apply changes requested by reviewer

apardyl reviewed (approved):
code looks OK, please manually double check if it actually works as no tests are present.
